File: blind_75_leetcode/python/s0133_clone_graph.py
from typing import Dict
from copy import copy

# Definition for a Node.
class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Solution:
    def cloneGraph(self, node: "Node") -> "Node":
        if node is None:
            return None
        visited = {}
        return Solution.bfs(node, visited)

    def bfs(node: "Node", visited: Dict[int, "Node"]) -> "Node":
        if visited.get(node.val):
            return visited[node.val]
        val = copy(node.val)
        new_node = Node(val, None)
        visited[val] = new_node
        if node.neighbors is not None:
            neighbors = [Solution.bfs(child, visited) for child in node.neighbors]
            new_node.neighbors = neighbors
        return new_node

File: blind_75_leetcode/src/solution/s0133_clone_graph.rs
#![allow(dead_code)]

use std::collections::HashMap;
use std::rc::Rc;

pub struct Solution {}

impl Solution {
    pub fn clone_graph(node: Rc<Node>) -> Rc<Node> {
        let node = Rc::new(node);
        let mut visited = HashMap::new();
        Solution::dfs(&node, &mut visited)
    }

    fn dfs(node: &Rc<Node>, visited: &mut HashMap<i32, Rc<Node>>) -> Rc<Node> {
        if let Some(node) = visited.get(&node.val) {
            return node.clone();
        }
        let val = node.val.clone();
        let neighbors = node
            .neighbors
            .iter()
            .map(|n| Solution::dfs(n, visited))
            .collect();
        let node = Rc::new(Node::new(val, neighbors));
        visited.insert(val, node.clone());
        node
    }
}

pub struct Node {
    val: i32,
    neighbors: Vec<Rc<Node>>,
}

impl Node {
    pub fn new(val: i32, neighbors: Vec<Rc<Node>>) -> Self {
        Node { val, neighbors }
    }
}

#[cfg(test)]
mod tests_133 {
    #[allow(unused_imports)]
    use super::*;

    #[test]
    fn it_works() {
        assert_eq!(0, 0);
    }
}
